crates/js-sys/src/lib.rs

```diff
@@ -4978,6 +4978,25 @@ macro_rules! arrays {
                 self.raw_copy_to(dst);
             }
 
+            /// Copy the contents of this JS typed array into the raw destination
+            /// memory, that might be unitialized.
+            ///
+            /// This function will efficiently copy the memory from a typed
+            /// array into this wasm module's own linear memory.
+            ///
+            /// # Panics
+            ///
+            /// This function will panic if this typed array's length is
+            /// different than the length argument passed into the function.
+            pub fn copy_to_mem_raw(&self, dst: *mut $ty, length: usize) {
```

Comment:
First of all, this function should be marked as `unsafe` since it's writing to a raw pointer.

Second, it seems to pretty much duplicate `raw_copy_to` / `copy_to` above, except by accepting raw pointer instead of a slice. Thus, its body could be replaced with just

```rust
self.copy_to(std::slice::from_raw_parts_mut(dst, length))
```

at which point it's not very clear to me what value does it bring as a separate API.

Reply:
@RReverser

> First of all, this function should be marked as `unsafe` since it's writing to a raw pointer.

Agreed.

> Second, it seems to pretty much duplicate `raw_copy_to` / `copy_to` above, except by accepting raw pointer instead of a slice.

I brought this up in here: https://github.com/rustwasm/wasm-bindgen/pull/1855#discussion_r422647554

In any case, let me know what you want to do with this PR. `view_mut_raw`  has already been merged, so we can close the issue. This PR is a follow-up that started with idea of introducing a new `copy_to_xxx` that works with `MaybeUninit` and ended up with raw pointer.

Do you think it is worth converting the test cases to use `copy_to`? There is only one existing `copy_to` test case, it tests `Int32Array`.

Reply:
> Second, it seems to pretty much duplicate `raw_copy_to` / `copy_to` above, except by accepting raw pointer instead of a slice. Thus, its body could be replaced with just
> 
> ```rust
> self.copy_to(std::slice::from_raw_parts_mut(dst, length))
> ```
> 
> at which point it's not very clear to me what value does it bring as a separate API.

No. Constructing a reference to uninitialized memory is illegal and undefined behavior. The entire point of this issue was to be able to work with uninitialized memory in a way that is performant and doesn't invoke undefined behavior.

I'm no longer using wasm-bindgen, so I no longer have any vested issue in the outcome of this PR. `view_mut_raw` probably does the job but I haven't verified this. I think it didn't exist when this issue was opened.

Reply:
> In any case, let me know what you want to do with this PR.

Oh, sorry, I'm not representative of the team, just someone fairly actively involved with Wasm and Wasm-bindgen. It's up to them anyway :)

> No. Constructing a reference to uninitialized memory is illegal and undefined behavior.

Right, that's a good point - the destination should be `&mut [MaybeUninit<$ty>]` instead. Maybe worth adding a function using that accepts such uninitialized slice instead, and then switching `copy_to` to use it under the hood?

Reply:
@That3Percent

The undefined behavior is pretty well document in here https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html

If you pass a pointer to aligned, continuous block of memory of proper size you are good :). The whole thing is unsafe. The caller should understand what they are doing.

I agree that `MaybeUninit` is the way to go for one reason, it is `safe`... however ` MaybeUninit::uninit_array()` is still nightly-only experimental API:
https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.uninit_array
So to allocate uninitialized array, callers will have to use `unsafe` Rust. So one way or another it has to be unsafe.

This issue should be reviewed again when `MaybeUninit::uninit_array()` is stabilized.

Reply:
Hold up. No, this is still undefined behavior. It is not sufficient that the pointer is aligned, contain a continuous block of memory, etc. These are merely additional constraints on top the normal invariants that must be upheld for references (of which there are many!)

Please see the [documentation ](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html)on `MaybeUninit` which states:

> This makes it undefined behavior to have uninitialized data in a variable even if that variable has an integer type, which otherwise can hold any fixed bit pattern

Also please see the [section in the nomicon](https://doc.rust-lang.org/nomicon/unchecked-uninit.html) about unchecked uninitialized memory which states:

> It is illegal to construct a reference to uninitialized data

Saying "the whole thing is unsafe" is not a license to be fast and loose. "Unsafe" means that you are providing a guarantee to the compiler that there is no undefined behavior. But this is clearly undefined behavior and illegal. I'm not sure how I can state this more strongly.

Reply:
Furthermore, the API `MaybeUninit::uninit_array` is irrelevant to the discussion. It is for creating an uninitialized array, not a slice.

Creating an uninitialized slice is straightforward and does not require any nightly APIs. Here is a link to the relevant [docs](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#initializing-an-array-element-by-element)

Reply:
FWIW another alternative is to accept a pointer-based slice: `*mut [T]`. It can be constructed safely via [`std::ptr::slice::from_raw_parts`](https://doc.rust-lang.org/stable/std/ptr/fn.slice_from_raw_parts.html) - since it's a raw pointer, it's not a subject to the same UB, and can be even cast from either a regular slice or `MaybeUninit` one.